// @revolist/angular-datagrid/esm2022/lib/editor.adapter.mjs
import { TemplateConstructor } from './renderer';
export class EditorAdapter {
    constructor(EditorComponent, injector, column, save, close) {
        this.EditorComponent = EditorComponent;
        this.injector = injector;
        this.column = column;
        this.save = save;
        this.close = close;
        this.element = null;
        this.renderedComponent = null;
    }
    // optional, called after editor rendered
    componentDidRender() { }
    // optional, called after editor destroyed
    disconnectedCallback() {
        this.renderedComponent?.destroy();
        this.renderedComponent = null;
    }
    render(h) {
        return h('span', {
            key: `${this.column.prop}-${this.editCell?.rowIndex || 0}`,
            ref: (el) => this.renderedComponent = TemplateConstructor(el, this.EditorComponent, {
                ...this.editCell,
                column: this.column,
                save: this.save,
                close: this.close,
            }, this.injector),
        });
    }
}
